scripts/parser.py

```diff
@@ -0,0 +1,433 @@
+        )
+
+    @staticmethod
+    def parse_fastqs_structure(fastqs) -> List[List[str]]:
+        """
+        Takes a list of fastqs, and a set of nested lists of each R1 + R2 read.
+
+        >>> GenericParser.parse_fastqs_structure(['20210727_PROJECT1_L002_R2.fastq.gz', '20210727_PROJECT1_L002_R1.fastq.gz', '20210727_PROJECT1_L001_R2.fastq.gz', '20210727_PROJECT1_L001_R1.fastq.gz'])
+        [['20210727_PROJECT1_L002_R2.fastq.gz', '20210727_PROJECT1_L002_R1.fastq.gz'], ['20210727_PROJECT1_L001_R2.fastq.gz', '20210727_PROJECT1_L001_R1.fastq.gz']]
+
+
+        """
+        # find last instance of R\d, and then group by prefix on that
+        sorted_fastqs = sorted(fastqs)
+        r_matches = {r: rmatch.search(r) for r in sorted_fastqs}
+        no_r_match = [r for r, matched in r_matches.items() if matched is None]
+        if no_r_match:
+            no_r_match_str = ', '.join(no_r_match)
+            raise ValueError(
+                f"Couldn't detect the format of FASTQs (expected match for regex '{rmatch.pattern}'): {no_r_match_str}"
+            )
+
+        values = []
+        for _, grouped in groupby(sorted_fastqs, lambda r: r[: r_matches[r].start()]):
+            values.append(sorted(grouped))
+
+        return values
+
+    def create_file_object(
+        self,
+        filename: str,
+        secondary_files: List[Dict[str, any]] = None,
+    ) -> Dict[str, any]:
+        """Takes filename, returns formed CWL dictionary"""
+        checksum = None
+        md5_filename = filename + '.md5'
+        if self.file_exists(md5_filename):
+            contents = self.file_contents(md5_filename)
+            if contents:
+                checksum = f'md5:{contents.strip()}'
+
+        d = {
+            'location': self.file_path(filename),
+            'basename': os.path.basename(filename),
+            'class': 'File',
+            'checksum': checksum,
+            'size': self.file_size(filename),
+        }
+
+        if secondary_files:
+            d['secondaryFiles'] = secondary_files
+
+        return d
+
+    def create_secondary_file_objects_by_potential_pattern(
+        self, filename, potential_secondary_patterns: List[str]
+    ) -> List[Dict[str, any]]:
+        """
+        Take a base filename and potential secondary patterns:
+        - Try each secondary pattern, see if it works
+        - If it works, create a CWL file object
+        - return a list of those secondary file objects that exist
+        """
+        secondaries = []
+        for sec in potential_secondary_patterns:
+            sec_file = apply_secondary_file_format_to_filename(filename, sec)
+            if self.file_exists(sec_file):
+                secondaries.append(self.create_file_object(sec_file))
+
+        return secondaries
+
+
+def apply_secondary_file_format_to_filename(
+    filepath: Optional[str], secondary_file: str
+):
+    """
+    You can trust this function to do what you want
```

Comment:
Just a minor comment, I usually try to prefix internal functions with an underscore: `_apply_secondary_file_format_to_filename`, in which case linter doesn't require a docstring. But if a function is meant to be imported, better to add a meaningful docstring.